### PR TITLE
8041488: Locale-Dependent List Patterns

### PR DESCRIPTION
**Pull request by naotoj**
Introducing a new formatting class for locale-dependent list patterns. The class is to provide the functionality from the Unicode Consortium's LDML specification for [list patterns](https://www.unicode.org/reports/tr35/tr35-general.html#ListPatterns). For example, given a list of String as "Monday", "Wednesday", "Friday", its `format` method would produce "Monday, Wednesday, and Friday" in US English. A CSR has also been drafted, and its draft javadoc can be viewed here: https://cr.openjdk.org/~naoto/JDK-8041488-ListPatterns-PR/api.00/java.base/java/text/ListFormat.html

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8295240](https://bugs.openjdk.org/browse/JDK-8295240) to be approved

### Issues
 * [JDK-8041488](https://bugs.openjdk.org/browse/JDK-8041488): Locale-Dependent List Patterns (**Enhancement** - P4)
 * [JDK-8295240](https://bugs.openjdk.org/browse/JDK-8295240): Locale-Dependent List Patterns (**CSR**)

### Reviewers
 * [Joe Wang](https://openjdk.org/census#joehw) (@JoeWang-Java - **Reviewer**) ⚠️ Review applies to [316ed12d](https://git.openjdk.org/jdk/pull/15130/files/316ed12d09ff64410465f8f2e72924d54c7738c5)
 * [Roger Riggs](https://openjdk.org/census#rriggs) (@RogerRiggs - **Reviewer**) ⚠️ Review applies to [316ed12d](https://git.openjdk.org/jdk/pull/15130/files/316ed12d09ff64410465f8f2e72924d54c7738c5)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15130/head:pull/15130` \
`$ git checkout pull/15130`

Update a local copy of the PR: \
`$ git checkout pull/15130` \
`$ git pull https://git.openjdk.org/jdk.git pull/15130/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15130`

View PR using the GUI difftool: \
`$ git pr show -t 15130`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15130.diff">https://git.openjdk.org/jdk/pull/15130.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15130#issuecomment-1663094305)

### CODE REVIEW

File: src/java.base/share/classes/java/text/ListFormat.java [hunk 1]
```diff
@@ -0,0 +1,571 @@
...
+    public static ListFormat getInstance() {
+        return getInstance(Locale.getDefault(Locale.Category.FORMAT), Type.STANDARD, Style.FULL);
+    }
+
+    /**
+     * {@return the list format object for the specified {@code Locale}, {@code Type},
+     * and {@code Style}}
+     * @param locale Locale to be used, not null
+     * @param type type of the list format. One of STANDARD/OR/UNIT, not null
+     * @param style style of the list format. One of FULL/SHORT/NARROW, not null
```
Comment (RogerRiggs) [inline src/java.base/share/classes/java/text/ListFormat.java:199]: It would be ease navigation in the javadoc if there was a link to the ListFormat.Type and ListFormat.Style classes, perhaps Locale also.
Use the normal prose form STANDARD, OR, or UNIT, instead of the "/" shorthand. Here and throughout the javadoc.  Ditto FULL, SHORT, or NARROW.
Also use Locale, Type, and Style instead of the shorthand.

File: src/java.base/share/classes/java/text/ListFormat.java [hunk 2]
```diff
@@ -0,0 +1,571 @@
...
+     * <blockquote><pre>
+     * start := (start_before){0}start_between{1}
+     * middle := {0}middle_between{1}
+     * end := {0}end_between{1}(end_after)
+     * two := (two_before){0}two_between{1}(two_after)
+     * three := (three_before){0}three_between{1}three_between{2}(three_after)
+     * </pre></blockquote>
+     * If parsing of the pattern string for start/middle/end fails, it throws an
+     * {@code IllegalArgumentException}. If two/three pattern string is empty, or
+     * fails on parsing, it falls back to
```
Comment (RogerRiggs) [inline src/java.base/share/classes/java/text/ListFormat.java:234]: I don't see code to try to parse the TWO or THREE patterns, where it could detect incorrectly formatted (user supplied) patterns.
If I understand correctly, a bad TWO or THREE pattern would not be detected until it is used in format or parse.

File: src/java.base/share/classes/java/text/ListFormat.java [hunk 3]
```diff
@@ -0,0 +1,571 @@
...
+        return
+            """
+            ListFormat [locale: "%s", start: "%s", middle: "%s",  end: "%s", two: "%s", three: "%s"]
+            """.formatted(locale.getDisplayName(), patterns[START], patterns[MIDDLE], patterns[END], patterns[TWO], patterns[THREE]);
+    }
+
+    private MessageFormat generateMessageFormat(Object[] input) {
+        var len = input.length;
+        return switch (len) {
+            case 0 -> throw new IllegalArgumentException("There should at least be one input string");
```
Comment (RogerRiggs) [inline src/java.base/share/classes/java/text/ListFormat.java:480]: It is unfortunate that LDML did not consider the case of an empty list. For example, to supply "NONE" or "n/a" or similar.  The developer will need to deal with that separately.

File: src/java.base/share/classes/java/text/ListFormat.java [hunk 4]
```diff
@@ -0,0 +1,571 @@
...
+        Objects.requireNonNull(type);
+        Objects.requireNonNull(style);
+        return new ListFormat(locale, LocaleProviderAdapter.forType(LocaleProviderAdapter.Type.CLDR)
+                .getLocaleResources(locale)
+                .getListPatterns(type, style));
+    }
+
+    /**
+     * {@return the list format for the specified patterns}
+     * <p>
```
Comment (RogerRiggs) [inline src/java.base/share/classes/java/text/ListFormat.java:213]: LDML includes a section on the "Gender of Lists";  If later support was desired, is this API extensible to include that?  https://www.unicode.org/reports/tr35/tr35-general.html#List_Gender

File: src/java.base/share/classes/java/text/ListFormat.java [hunk 5]
```diff
@@ -0,0 +1,571 @@
...
+    /**
+     * {@return the parsed list of Strings from the {@code source} String}
+     *
+     * Note that {@link #format(List)} and this method
+     * may not guarantee a round-trip, if the input strings contain ambiguous
+     * delimiters. For example, a String list {@code "a, b,", "c"} will be
+     * formatted as {@code "a, b, and c"}, but may be parsed as
+     * {@code "a", "b", "c"}.
+     *
+     * @param source the String to parse, not null.
```
Comment (RogerRiggs) [inline src/java.base/share/classes/java/text/ListFormat.java:350]: String does not need to be capitalized.

File: src/java.base/share/classes/java/text/ListFormat.java [hunk 6]
```diff
@@ -0,0 +1,571 @@
...
+        if (parseObject(source, pp) instanceof List<?> orig) {
+            // parseObject() should've returned List<String>
+            return orig.stream().map(o -> (String)o).toList();
+        } else {
+            throw new ParseException("Parse failed", pp.getErrorIndex());
+        }
+    }
+
+    /**
+     * Parses text from a string to produce a list of {@code String}s.
```
Comment (RogerRiggs) [inline src/java.base/share/classes/java/text/ListFormat.java:365]: No need to capitalize String if list is not capitalized. (or be consistent)

File: src/java.base/share/classes/java/text/ListFormat.java [hunk 7]
```diff
@@ -0,0 +1,571 @@
...
+        if (parsed == null) {
+            parsed = new String[]{source.substring(parsePos.index)};
+            parsePos.setIndex(source.length());
+        }
+
+        if (parsed instanceof Object[] objs) {
+            parsePos.setErrorIndex(-1);
+            return Arrays.asList(objs);
+        }
+        throw new InternalError("MessageFormat.parseObject() should return Object[]");
```
Comment (RogerRiggs) [inline src/java.base/share/classes/java/text/ListFormat.java:428]: The error message is inconsistent with line 426, returning a List<String>.

File: src/java.base/share/classes/java/text/ListFormat.java [hunk 8]
```diff
@@ -0,0 +1,571 @@
...
+        if (arguments instanceof List<?> objs) {
+            var a = objs.toArray(new Object[0]);
+            return generateMessageFormat(a).formatToCharacterIterator(a);
+        } else {
+            throw new IllegalArgumentException("The arguments should be an Object list");
+        }
+    }
+
+    /**
+     * {@inheritDoc}
```
Comment (RogerRiggs) [inline src/java.base/share/classes/java/text/ListFormat.java:442]: The javadoc from Object.equals doesn't give any information about what constituted equals for a ListFormat.
Namely, that the locale and patterns are equal.

File: src/java.base/share/classes/java/text/ListFormat.java (hunk 2 above)
Comment (naotoj) [reply src/java.base/share/classes/java/text/ListFormat.java:234]: Right. Will provide validation for them

File: src/java.base/share/classes/java/text/ListFormat.java (hunk 4 above)
Comment (naotoj) [reply src/java.base/share/classes/java/text/ListFormat.java:213]: IIUC, although it is described within "list patterns" in LDML, it is orthogonal to "list patterns." The purpose of the "Gender of Lists" seems to me the resulting gender of the whole list of names, e.g., "Tom and Mary" is "other" in English, but it is "male" in Arabic (I just used the sample XML in the spec). So I believe it is not an extension of the class.

File: src/java.base/share/classes/java/text/ListFormat.java (hunk 8 above)
Comment (naotoj) [reply src/java.base/share/classes/java/text/ListFormat.java:442]: Will add those ListFormat specific considerations.

File: src/java.base/share/classes/java/text/ListFormat.java [hunk 9]
```diff
@@ -0,0 +1,591 @@
...
+     * @throws NullPointerException if {@code source} or {@code parsePos} is null.
+     */
+    @Override
+    public Object parseObject(String source, ParsePosition parsePos) {
+        Objects.requireNonNull(source);
+        Objects.requireNonNull(parsePos);
+        var sm = startPattern.matcher(source);
+        var em = endPattern.matcher(source);
+        Object parsed = null;
+        if (sm.find(parsePos.index) && em.find(parsePos.index)) {
```
Comment (JoeWang-Java) [inline src/java.base/share/classes/java/text/ListFormat.java:408]: Would it be better to call getIndex() instead? (same below)

File: src/java.base/share/classes/java/text/ListFormat.java [hunk 10]
```diff
@@ -0,0 +1,591 @@
...
+ * {@link Type#OR OR}, and {@link Type#UNIT UNIT}, also three styles for each
+ * type are provided: {@link Style#FULL FULL}, {@link Style#SHORT SHORT}, and
+ * {@link Style#NARROW NARROW}. The following snippet is an example of formatting
+ * the list of Strings {@code "Foo", "Bar", "Baz"} in US English with
+ * {@code STANDARD} type and {@code FULL} style:
+ * {@snippet lang=java :
+ * ListFormat.getInstance(Locale.US, ListFormat.Type.STANDARD, ListFormat.Style.FULL)
+ *     .format(List.of("Foo", "Bar", "Baz"))
+ * }
+ * This will produce the concatenated list string, "Foo, Bar, and Baz" as seen in
```
Comment (JoeWang-Java) [inline src/java.base/share/classes/java/text/ListFormat.java:58]: With this sample code, if the Style is changed to SHORT, it would produce the same string. Would it be better to use the weekdays instead of Foo, Bar and Baz (as in the Unicode spec)? Esp. with the UNIT type, those examples explained it better, e.g. NARROW produces 3′ 7″.

Also, if the instance is of STANDARD/SHORT, does it format List.of("January", "February", "March") and return "Jan., Feb., and Mar.", or 3 feet, 7 inches to 3 ft, 7 in? The format method states simply "Returns the string that consists of the input strings, concatenated with the patterns of this ListFormat."  I wonder if it'd be helpful to explain a bit more or add one more sample.

File: src/java.base/share/classes/java/text/ListFormat.java [hunk 11]
```diff
@@ -0,0 +1,591 @@
...
+ * <thead>
+ * <tr><th scope="col"></th>
+ *     <th scope="col">FULL</th>
+ *     <th scope="col">SHORT</th>
+ *     <th scope="col">NARROW</th></tr>
+ * </thead>
+ * <tbody>
+ * <tr><th scope="row" style="text-align:left">STANDARD</th>
+ *     <td>Foo, Bar, and Baz</td>
+ *     <td>Foo, Bar, &amp; Baz</td>
```
Comment (JoeWang-Java) [inline src/java.base/share/classes/java/text/ListFormat.java:71]: Is "&amp;" a typo?  It's still "and" in the Unicode spec's "standard-short" format, e.g. "Jan., Feb., and Mar."

File: test/jdk/java/text/Format/ListFormat/TestListFormat.java [hunk 12]
```diff
@@ -0,0 +1,262 @@
...
+                arguments(Locale.US, ListFormat.Type.STANDARD, ListFormat.Style.SHORT,
+                        "foo, bar, & baz", true),
+                arguments(Locale.US, ListFormat.Type.OR, ListFormat.Style.SHORT,
+                        "foo, bar, or baz", true),
+                arguments(Locale.US, ListFormat.Type.UNIT, ListFormat.Style.SHORT,
+                        "foo, bar, baz", true),
+                arguments(Locale.US, ListFormat.Type.STANDARD, ListFormat.Style.NARROW,
+                        "foo, bar, baz", true),
+                arguments(Locale.US, ListFormat.Type.OR, ListFormat.Style.NARROW,
+                        "foo, bar, or baz", true),
```
Comment (JoeWang-Java) [inline test/jdk/java/text/Format/ListFormat/TestListFormat.java:157]: Same as in the ListFormat class, the expected results are the same "foo, bar, or baz" when different Styles are specified.

File: src/java.base/share/classes/java/text/ListFormat.java (hunk 10 above)
Comment (naotoj) [reply src/java.base/share/classes/java/text/ListFormat.java:58]: In fact, the sample in LDML's page seems to be incorrect. `standard-short` in English is defined as:
```
		<listPattern type="standard-short">
			<listPatternPart type="start">{0}, {1}</listPatternPart>
			<listPatternPart type="middle">{0}, {1}</listPatternPart>
			<listPatternPart type="end">{0}, &amp; {1}</listPatternPart>
			<listPatternPart type="2">{0} &amp; {1}</listPatternPart>
		</listPattern>
```
in `en.xml` file. So `&` is expected rather than `and`.

> Also, if the instance is of STANDARD/SHORT, does it format List.of("January", "February", "March") and return "Jan., Feb., and Mar.", or 3 feet, 7 inches to 3 ft, 7 in?

No it does not. The `format()` method does not alter the passed input strings, so it would not convert "January" to "Jan." even if `SHORT` Style is specified. I have added some extra explanations, that those patterns vary depending on the locale providers.

File: src/java.base/share/classes/java/text/ListFormat.java (hunk 11 above)
Comment (naotoj) [reply src/java.base/share/classes/java/text/ListFormat.java:71]: Again `ampersand` is the correct pattern for `SHORT` in English.

File: test/jdk/java/text/Format/ListFormat/TestListFormat.java (hunk 12 above)
Comment (naotoj) [reply test/jdk/java/text/Format/ListFormat/TestListFormat.java:157]: Yes, those are exactly what are defined in CLDR. (They could have chosen `|` for SHORT style, but that would be not so common in plain English I guess)

File: src/java.base/share/classes/java/text/ListFormat.java (hunk 10 above)
Comment (JoeWang-Java) [reply src/java.base/share/classes/java/text/ListFormat.java:58]: Ah, I see. Those samples did give an impression that the format does more than concatenation.

File: test/jdk/java/text/Format/ListFormat/TestListFormat.java [hunk 13]
```diff
@@ -0,0 +1,262 @@
...
+        assertEquals(-1, pp.getErrorIndex());
+    }
+
+    private static void compareResult(ListFormat f, List<String> input, String expected, boolean roundTrip) throws ParseException {
+        var result = f.format(input);
+        assertEquals(expected, result);
+        if (roundTrip) {
+            assertEquals(input, f.parse(result));
+        }
+    }
```
Comment (RogerRiggs) [inline test/jdk/java/text/Format/ListFormat/TestListFormat.java:261]: There should be a test of `formatToCharacterIterator` method.

File: src/java.base/share/classes/java/text/ListFormat.java [hunk 14]
```diff
@@ -0,0 +1,593 @@
...
+    /**
+     * {@return the available locales that support list formatting}
+     */
+    public static Locale[] getAvailableLocales() {
+        // Same as a typical format class
+        return DateFormat.getAvailableLocales();
+    }
+
+    /**
+     * {@return the list format object for the default
```
Comment (RogerRiggs) [inline src/java.base/share/classes/java/text/ListFormat.java:199]: "list format" will read better using the class name `ListFormat`. Here an in other `getInstance` methods.
As is, in prose it looks like it returns some kind of list.

**Comment by naotoj** [comment]
Thanks, Roger. All comments make sense to me. Will update the PR soon.

File: src/java.base/share/classes/java/text/ListFormat.java [hunk 15]
```diff
@@ -0,0 +1,594 @@
...
+import sun.util.locale.provider.LocaleProviderAdapter;
+
+/**
+ * {@code ListFormat} produces or parses a list of concatenated
+ * strings in a locale-sensitive way. Use {@code ListFormat} to construct a list
+ * of strings displayed for end users. For example, displaying 3 weekdays
+ * as "Monday, Wednesday, and Friday". This class provides the functionality
+ * defined in Unicode Consortium's LDML specification for
+ * <a href="https://www.unicode.org/reports/tr35/tr35-general.html#ListPatterns">
+ * List Patterns</a>.
```
Comment (JoeWang-Java) [inline src/java.base/share/classes/java/text/ListFormat.java:46]: The main function, it seems to me, is to change the representation from one form to another. what would you think about the following:

The {@code ListFormat} class is a tool for converting a list of strings to a text representation and vice versa in a locale-sensitive way. It transforms strings to text in accordance with the List Patterns (link) as defined in Unicode Consortium's LDML specification. For example, it can be used to format a list of 3 weekdays, i.e. "Monday", "Wednesday", "Friday", as "Monday, Wednesday, and Friday" in an inclusive list pattern.

File: src/java.base/share/classes/java/text/ListFormat.java [hunk 16]
```diff
@@ -0,0 +1,594 @@
...
+ * <p>
+ * Three types of concatenation are provided: {@link Type#STANDARD STANDARD},
```
Comment (JoeWang-Java) [inline src/java.base/share/classes/java/text/ListFormat.java:48]: A "Type" and "Style" together make up a specific pattern. It might be good to introduce the term "List Patterns" here first, that is, moving the introduction of patterns to the class description from the 1-arg getInstance method. once we have the terms established, we can then delve into the specific cases "types" and "styles" represent. Something like:

<h2>List Patterns</h2>
List Patterns are rules that define how a series or list is formed ... (include the description for the getInstance(String[] patterns) here)

<h2>Standard Patterns</h2>
{@code ListFormat} supports a few pre-defined patterns with a combination of Type (link) and Style(link). Types and Styles are defined as follows.

<h3>Type</h3>
{@link Type#STANDARD STANDARD}: a simple list with conjunction "and";

...

<h3>Style</h3>
{@link Style#FULL FULL}: uses the conjunction word such as "and";

{@link Style#SHORT SHORT}: uses the shorthand of the conjunction word, "&" (ampersand) for "and" for example;

{@link Style#NARROW NARROW}: uses no conjunction word.

For example, a combination of {@link Type#STANDARD STANDARD} and {@link Style#FULL FULL} forms an inclusive list pattern.

File: src/java.base/share/classes/java/text/ListFormat.java [hunk 17]
```diff
@@ -0,0 +1,594 @@
...
+        /**
+         * The {@code OR} ListFormat style. This style concatenates
+         * elements in "or" enumeration.
+         */
+        OR,
+
+        /**
+         * The {@code UNIT} ListFormat style. This style concatenates
+         * elements, useful for enumerating units.
+         */
```
Comment (JoeWang-Java) [inline src/java.base/share/classes/java/text/ListFormat.java:560]: The word "style" used in Type, I assume you meant "type"? Just that it might be confused with Style below.

Same as previous comments, a combination of Type and Style, if I understand correctly, forms a specific pattern. I might say something about it in the enum class description.

A STANDARD type then is a simple list with conjunction "and", or an inclusive list, and etc.

File: src/java.base/share/classes/java/text/ListFormat.java [hunk 18]
```diff
@@ -0,0 +1,594 @@
...
+            case 1 -> new MessageFormat("{0}", locale);
+            case 2, 3 -> new MessageFormat(patterns[len + 1], locale);
+            default -> new MessageFormat(createMessageFormatString(len), locale);
+        };
+    }
+
+    private String createMessageFormatString(int count) {
+        var sb = new StringBuilder(256).append(patterns[START]);
+        IntStream.range(2, count - 1).forEach(i -> sb.append(middleBetween).append("{").append(i).append("}"));
+        sb.append(patterns[END].replaceFirst("\\{0}", "").replaceFirst("\\{1}", "\\{" + (count - 1) + "\\}"));
```
Comment (JoeWang-Java) [inline src/java.base/share/classes/java/text/ListFormat.java:521]: From what it looks, it could be a concern for potentially adding large number of long strings with a list of small items. I don't seem to see where the input is limited.

File: src/java.base/share/classes/java/text/ListFormat.java (hunk 15 above)
Comment (naotoj) [reply src/java.base/share/classes/java/text/ListFormat.java:46]: Thanks. Will modify the wording in the next revision. I think we should stick to the wording `format`/`parse` here.

File: src/java.base/share/classes/java/text/ListFormat.java (hunk 16 above)
Comment (naotoj) [reply src/java.base/share/classes/java/text/ListFormat.java:48]: I think that Type/Style/Locale forming a specific pattern is an implementation detail, so I would not describe it in the spec (although as you say they form a specific pattern in the impl). It could be that an impl of 3-arg getInstance() can be independent of patterns described in 1-arg getInstance().

File: src/java.base/share/classes/java/text/ListFormat.java (hunk 17 above)
Comment (naotoj) [reply src/java.base/share/classes/java/text/ListFormat.java:560]: Ah, good catch. Will correct those style/type typos.

> A STANDARD type then is a simple list with conjunction "and", or an inclusive list, and etc.

I could not quite catch what you meant by this. Can you please elaborate on it more?

File: src/java.base/share/classes/java/text/ListFormat.java (hunk 18 above)
Comment (naotoj) [reply src/java.base/share/classes/java/text/ListFormat.java:521]: Good point. Will add some kind of limitation.

File: src/java.base/share/classes/java/text/ListFormat.java (hunk 15 above)
Comment (RogerRiggs) [reply src/java.base/share/classes/java/text/ListFormat.java:46]: I'd stick to the general first sentence structure used by DateFormat, Format, and MessageFormat.
("tools" in OpenJDK are standalone programs.)

For example, 
"ListFormat formats and parses lists of strings."

File: src/java.base/share/classes/java/text/ListFormat.java (hunk 16 above)
Comment (RogerRiggs) [reply src/java.base/share/classes/java/text/ListFormat.java:48]: Type and Style select a pattern defined for the locale. (for the zero arg and three arg getInstance methods).
I think I would avoid the term "concatenation", the type defines a pattern that contains punctuation and connecting words. The style selects among the full, short, and narrow patterns.

File: src/java.base/share/classes/java/text/ListFormat.java (hunk 15 above)
Comment (naotoj) [reply src/java.base/share/classes/java/text/ListFormat.java:46]: I agree, Roger. The current version is derived from MessageFormat, which is the closest of all the *Format classes. Will come up with a better one.

Comment (JoeWang-Java) [reply src/java.base/share/classes/java/text/ListFormat.java:46]: Sounds good. I mainly want to say that ListFormat doesn't create (produce) new items but change (convert) the presentation from one form to another.

File: src/java.base/share/classes/java/text/ListFormat.java (hunk 18 above)
Comment (RogerRiggs) [reply src/java.base/share/classes/java/text/ListFormat.java:521]: I can't see what arbitrary limit would make sense.
But perhaps an APINote that  formatting the string from an excessively long list may exceed memory or string sizes. (Without being specific about the exception being thrown).

File: src/java.base/share/classes/java/text/ListFormat.java (hunk 16 above)
Comment (JoeWang-Java) [reply src/java.base/share/classes/java/text/ListFormat.java:48]: I thought they're already a part of the public API, and already described in the current javadoc, e.g. ListFormat.Style and ListFormat.Type (https://cr.openjdk.org/~naoto/JDK-8041488-ListPatterns-PR/api.00/java.base/java/text/ListFormat.html). I thought you meant to support general patterns (the List Patterns) and a few (specific) "type attributes" as in the LDML spec.

Comment (naotoj) [reply src/java.base/share/classes/java/text/ListFormat.java:48]: Yes, they are public API. My intention was to keep those patterns exposure as minimal as possible, as they are complicated so that users who just wish to acquire formats only with Type/Style/Locale need not know it. Only those who want to customize their own patterns can do so with the 1-arg getInstance(), thus I wanted the description in that method.

File: src/java.base/share/classes/java/text/ListFormat.java (hunk 18 above)
Comment (naotoj) [reply src/java.base/share/classes/java/text/ListFormat.java:521]: OK, will ad that @apiNote in the 1-arg getInstance()

File: src/java.base/share/classes/java/text/ListFormat.java (hunk 17 above)
Comment (JoeWang-Java) [reply src/java.base/share/classes/java/text/ListFormat.java:560]: In a List Pattern, Type, as in the current document, defined its type as in the language, while Style the writing style. I guess I would try to avoid stating it "concatenates elements", but rather represents a simple list with conjunction "and", that is also called an inclusive list as in the language term.

File: src/java.base/share/classes/java/text/ListFormat.java [hunk 19]
```diff
@@ -0,0 +1,594 @@
...
+ * type are provided: {@link Style#FULL FULL}, {@link Style#SHORT SHORT}, and
+ * {@link Style#NARROW NARROW}. The following snippet is an example of formatting
+ * the list of Strings {@code "Foo", "Bar", "Baz"} in US English with
+ * {@code STANDARD} type and {@code FULL} style:
+ * {@snippet lang=java :
+ * ListFormat.getInstance(Locale.US, ListFormat.Type.STANDARD, ListFormat.Style.FULL)
+ *     .format(List.of("Foo", "Bar", "Baz"))
+ * }
+ * This will produce the concatenated list string, "Foo, Bar, and Baz" as seen in
+ * the following table:
```
Comment (RogerRiggs) [inline src/java.base/share/classes/java/text/ListFormat.java:59]:
```suggestion
 * the following:
```

File: src/java.base/share/classes/java/text/ListFormat.java [hunk 20]
```diff
@@ -0,0 +1,594 @@
...
+ *     <td>Foo, Bar, or Baz</td>
+ *     <td>Foo, Bar, or Baz</td>
+ * <tr><th scope="row" style="text-align:left">UNIT</th>
+ *     <td>Foo, Bar, Baz</td>
+ *     <td>Foo, Bar, Baz</td>
+ *     <td>Foo Bar Baz</td>
+ * </tbody>
+ * </table>
+ * Note that the above chart represents the examples in the CLDR definition,
+ * and there could be different patterns from other locale providers.
```
Comment (RogerRiggs) [inline src/java.base/share/classes/java/text/ListFormat.java:84]:
```suggestion
 * Note: these examples are from CLDR, there could be different patterns from other locale providers.
```

File: src/java.base/share/classes/java/text/ListFormat.java [hunk 21]
```diff
@@ -0,0 +1,594 @@
...
+ * <p>
+ * Alternatively, Locale, Type, and/or Style independent instances
+ * can be created with {@link #getInstance(String[])}. The String array to the
+ * method specifies the delimiting patterns for the start/middle/end portion of
+ * the formatted string, as well as optional specialized patterns for two or three
+ * elements. Refer to the method description for more detail.
```
Comment (RogerRiggs) [inline src/java.base/share/classes/java/text/ListFormat.java:90]:
```suggestion
 * Alternatively, more flexible patterns can be constructed from the pattern parts for the start, middle, 
 * and end of the formatted list as well as specialized patterns for lists of two or three elements.
 * Refer to the {@link #getInstance(String[])} method description for more detail.
```

Comment (naotoj) [reply src/java.base/share/classes/java/text/ListFormat.java:89]: For this part, I would not use `patterns` in the first suggested sentence but use `more flexible ListFormat instances` instead.

**Review by JoeWang-Java — APPROVED** [review]

File: src/java.base/share/classes/java/text/ListFormat.java [hunk 22]
```diff
@@ -0,0 +1,600 @@
...
+    /**
+     * Formats an object and appends the resulting text to a given string
+     * buffer. The object should either be a List or an array of Objects.
+     *
+     * @apiNote Formatting the string from an excessively long list or array
+     *          may exceed memory or string sizes.
+     * @param obj    The object to format. Must be a List or an array
+     *               of Object.
+     * @param toAppendTo    where the text is to be appended
+     * @param pos    Ignored. Not used in ListFormat. May be null
```
Comment (RogerRiggs) [inline src/java.base/share/classes/java/text/ListFormat.java:345]: Curious, why not used? 
I could see a use to identity the string inserted to enable highlighting or other markup around the new string.

**Review by RogerRiggs — APPROVED** [review]
Looking very good.

File: src/java.base/share/classes/java/text/ListFormat.java (hunk 22 above)
Comment (naotoj) [reply src/java.base/share/classes/java/text/ListFormat.java:345]: `FieldPosition` is dedicated to identifying fields in *Format classes which are either `Format.Field` or fields that have names ending with "_FIELD", which ListFormat has neither of them.

File: src/java.base/share/classes/java/text/ListFormat.java [hunk 23]
```diff
@@ -0,0 +1,600 @@
...
+ * <p>
+ * Alternatively, Locale, Type, and/or Style independent instances
+ * can be created with {@link #getInstance(String[])}. The String array to the
+ * method specifies the delimiting patterns for the start/middle/end portion of
+ * the formatted string, as well as optional specialized patterns for two or three
+ * elements. Refer to the method description for more detail.
+ * <p>
+ * On parsing, if some ambiguity is found in the input string, such as delimiting
+ * sequences being found in the input string, may produce the result that when formatted is not a
+ * round-trip with the corresponding formatting. For example, a two element String list
```
[the remark this answers is not in the document]
Comment (naotoj) [reply src/java.base/share/classes/java/text/ListFormat.java]: Thanks, Roger. Incorporated.

File: src/java.base/share/classes/java/text/ListFormat.java [hunk 24]
```diff
@@ -0,0 +1,600 @@
...
+        return sb.toString();
+    }
+
+    @java.io.Serial
+    private void readObject(ObjectInputStream stream)
+            throws IOException, ClassNotFoundException {
+        stream.defaultReadObject();
+        try {
+            init();
+        } catch (IllegalArgumentException iae) {
```
Comment (RogerRiggs) [inline src/java.base/share/classes/java/text/ListFormat.java:536]: If the `patterns` array contains a null, perhaps due to corrupted stream contents, `init()` will throw NPE.
I don't recommend catching NPE here, but perhaps `init()` should check for nulls and throw IAE.

Comment (naotoj) [reply src/java.base/share/classes/java/text/ListFormat.java:536]: Actually, null checks for the `patterns` array elements should be done also for the 1-arg factory method. Will add the check in `init()` and the corresponding test.
